### PR TITLE
MyDSpace Search component refactoring

### LINKED ISSUE #1044: MyDSpace page should use the same components as other search results pages & should not have custom object swapping/reloading functionality
**Describe the bug**

The changes made in #978 to fix #721 were not ideal as they treat the MyDSpace page's search results differently from other search results listing.  These changes however were deemed necessary because the MyDSpace page doesn't use the same components as other search results...see @atarix83's [comment here](https://github.com/DSpace/dspace-angular/pull/978#issuecomment-791450504).  This resulted in a need to refresh individual components within an existing list of results.

Ideally, we would NOT be refreshing/swapping individual objects in a list, and instead update the list while ensuring you "follow" or remain with the updated item in that list.  One way to possibly achieve this is described by @artlowel in [this comment](https://github.com/DSpace/dspace-angular/pull/978#issuecomment-790800076).

Unfortunately, we were not able to easily solve this in 7.0 without a larger refactor (as described in related tickets below).  Therefore, this work may need to be completed in conjunction with those related tickets.

**Related work**

* Discovered in review of #978 (and fixing this ticket will likely require removing or refactoring the work of the prior PR)
* Relates directly to #773 
* Also seems related to #741 (as that involves updating/deleting objects in lists)

### PR DESCRIPTION
**Pull request by atarix83**
## References
* Fixes #1044

## Description
This PR introduce a refactoring of the search and mydspace components 

## Instructions for Reviewers
Test all functionality where the search functionality is involved

List of changes in this PR:
* Created a new search.module that contains all components involved in the search 
* Created `form.module` and `comcol.module` in order to resolve circular dependencies after creating the search.module. This has a good impact on main bundle size because they don't belong anymore to the shared.module. More in general we should take in account a refactoring of the shared.module because not always it has been used as the angular guidelines recommend.
* The search component has been refactored in order to be able to meet all the needs of the various search functions. In particular now it relies always on the discovery configuration, and retrieve all search options (filters, sorts etc...) always based on the configuration
* The mydspace page has been refactored in order to use the new search component

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Comment by tdonohue** [comment]
@atarix83 : Just a quick note, it appears you accidentally broke an e2e test in this PR.  It looks like the e2e tests can no longer switch the the grid layout -- it's possible the name or location of the button just changed.  In any case, we should be able to start reviewing while you get the test fixed.

File: src/assets/i18n/en.json5 [hunk 1]
```diff
@@ -4019,8 +4019,12 @@
 
 
 
+  "workspace.search.results.head": "Your submissions",
+
   "workflowAdmin.search.results.head": "Administer Workflow",
 
+  "workflow.search.results.head": "All tasks",
```
Comment (tdonohue) [inline src/assets/i18n/en.json5:4026]: I'd like to recommend we change the text of this label to say "Workflow tasks".  I've always found "All tasks" to be a confusing name, as it's unclear what it refers to. 

So, this PR might be a good opportunity to just change the wording here, as well as in the `mydspace.show.workflow` i18n key.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@atarix83 : I gave this a test today, and it **works great overall**!  I've tested MyDSpace search options (both workflow & workspace), header/normal search, & related Entity search (e.g. on a Person page searching their publication).  All of them still work well... I've only found **two issues** (and both seem minor).

1. On the MyDSpace page, the "Detail view" no longer works with pagination. If you click that option, the first item shows up in detail...but if you page to the next item, the detailed view is lost. This seems to work fine on `main` (and the demo site), so it's a bug specific to this PR.
2. As mentioned above, the `e2e` tests are failing when it tries to check the "grid view" of the search results.  I think it's just that the selector has changed slightly on this line: https://github.com/DSpace/dspace-angular/blob/main/cypress/integration/search-page.spec.ts#L56  (I think the `href` part should just be `/search?spc.page=1&view=grid` now, or something similar to that.)  I think this should be likely easy to fix, but if you hit issues, let me know.

Finally, I have a very minor inline comment below to request rewording of a label. But we could do this in a separate tiny PR if needed, it's not required.

**Review by artlowel — CHANGES_REQUESTED** [review]
Thanks @atarix83!

The code looks good, and it mostly works, except for in the `ds-dynamic-lookup-relation-search-tab` component. This is the one you get when adding a relationship from the submission or the edit item page.

It sends the wrong queries to the backend. As a result you always get 

>Search query is not valid. Please check Solr query syntax best practices for further information about this error.

File: src/assets/i18n/en.json5 (hunk 1 above)
Comment (atarix83) [reply src/assets/i18n/en.json5]: done

**Comment by atarix83** [comment]
@tdonohue 

detail view should work now, can you check it again

@artlowel 
> The code looks good, and it mostly works, except for in the ds-dynamic-lookup-relation-search-tab component. This is the one you get when adding a relationship from the submission or the edit item page.

thanks for pointing it out, I've forget to change it in order to use the refactored search-component. Now it is like this, could you check if it is ok?

**Review by tdonohue — APPROVED** [review]
👍 Thanks @atarix83 !  Retested this today and all my prior feedback has been addressed.  So, this looks good to me, as soon as @artlowel has a chance to re-review & approve.

File: src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/search-tab/dynamic-lookup-relation-search-tab.component.ts [hunk 2]
```diff
@@ -306,4 +293,16 @@ export class DsDynamicLookupRelationSearchTabComponent implements OnInit, OnDest
       this.subscription.unsubscribe();
     }
   }
+
+  onResultFound($event: SearchObjects<DSpaceObject>) {
+    console.log($event);
```
Comment (artlowel) [inline src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/search-tab/dynamic-lookup-relation-search-tab.component.ts:298]: Forgot a console.log

**Review by artlowel — COMMENTED** [review]
Thanks @atarix83!

The relationship modal seems to work mostly now. But if I open it in certain situations (e.g. when trying to add a publication to [Simmons, Cameron, when using api7.dspace.org](http://localhost:4000/entities/person/b1b2c768-bda1-448a-a073-fc541e8b24d9/edit/relationships)) I get an exception now:
```
VM4461:85 ERROR TypeError: t.find is not a function
    at e.getRelatedItem (dynamic-lookup-relation-search-tab.component.ts:282:23)
    at dynamic-lookup-relation-search-tab.component.ts:257:23
    at Array.map (<anonymous>)
    at u._next (dynamic-lookup-relation-search-tab.component.ts:247:41)
    at u.__tryOrUnsub (Subscriber.js:183:16)
    at u.next (Subscriber.js:122:22)
    at l._next (Subscriber.js:72:26)
    at l.next (Subscriber.js:49:18)
    at o._next (map.js:35:26)
    at o.next (Subscriber.js:49:18)
```
Which refers to [this line](https://github.com/DSpace/dspace-angular/blob/d7388bcfd86448d748e64d9d5149fb9d8b0909c7/src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/search-tab/dynamic-lookup-relation-search-tab.component.ts#L282)

In that modal the viewmode selector is also shown (it isn't on the current main branch) and selecting grid mode will crash the modal.

**Comment by atarix83** [comment]
@artlowel 

bug should be fixed now could you check ?

**Review by artlowel — APPROVED** [review]
That works.

Thanks @atarix83!

**Pull request merged**